core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,278 @@
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the <code>Key</code>
+     * @return this builder
+     */
+    Build<T> columnVisibility(final T columnVisibility);
+  }
+
+  private static abstract class AbstractKeyBuilder<T> implements ColumnFamilyStep<T>, ColumnQualifierStep<T>,
+      ColumnVisibilityStep<T> {
+
+    protected static final byte EMPTY_BYTES[] = new byte[0];
+
+    protected T row = null;
+    protected T columnFamily = null;
+    protected T columnQualifier = null;
+    protected T columnVisibility = null;
+    protected long timestamp = Long.MAX_VALUE;
+    protected boolean deleted = false;
+
+    final public ColumnFamilyStep<T> row(final T row) {
+      this.row = row;
+      return this;
+    }
+
+    @Override
+    final public ColumnQualifierStep<T> columnFamily(final T columnFamily) {
+      this.columnFamily = columnFamily;
+      return this;
+    }
+
+    @Override
+    final public ColumnVisibilityStep<T> columnQualifier(final T columnQualifier) {
+      this.columnQualifier = columnQualifier;
+      return this;
+    }
+
+    @Override
+    final public Build<T> columnVisibility(final T columnVisibility) {
+      this.columnVisibility = columnVisibility;
+      return this;
+    }
+
+    @Override
+    final public Build<T> timestamp(long timestamp) {
+      this.timestamp = timestamp;
+      return this;
+    }
+
+    @Override
+    public Build<T> deleted(boolean deleted) {
+      this.deleted = deleted;
+      return this;
+    }
+
+    @Override
+    public Key build() {
+      return this.build(true);
+    }
+
+    @Override
+    public Build<T> column(final T columnFamily, final T columnQualifier, final T columnVisibility) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier).columnVisibility(columnVisibility);
+    }
+
+    @Override
+    public ColumnVisibilityStep<T> column(final T columnFamily, final T columnQualifier) {
+      return this.columnFamily(columnFamily).columnQualifier(columnQualifier);
+    }
+  }
+
+  private static class TextKeyBuilder extends AbstractKeyBuilder<Text> {
+
+    private final Text EMPTY_TEXT = new Text();
+
+    @Override
+    public Key build(boolean copyBytes) {
+      Text columnFamily = (this.columnFamily == null) ? EMPTY_TEXT : this.columnFamily;
+      Text columnQualifier = (this.columnQualifier == null) ? EMPTY_TEXT : this.columnQualifier;
+      Text columnVisibility = (this.columnVisibility == null) ? EMPTY_TEXT : this.columnVisibility;
+      return new Key(row.getBytes(), 0, row.getLength(), columnFamily.getBytes(), 0, columnFamily.getLength(),
+          columnQualifier.getBytes(), 0, columnQualifier.getLength(), columnVisibility.getBytes(), 0,
+          columnVisibility.getLength(), timestamp, deleted, copyBytes);
+    }
+  }
+
+  private static class ByteArrayKeyBuilder extends AbstractKeyBuilder<byte[]> {
```

Comment:
You mean a single `ByteBuffer` as input and the offsets of the components or multiple `ByteBuffer`s, one for each component? In the second case it's easy, I can extract the underlying array from the `ByteBuffer` while in the first case I would need special methods that get the offset and length to be used for each components and I don't think it's really nice. Can you give me an example of how would you use a `ByteBuffer` as input to build a `Key`?